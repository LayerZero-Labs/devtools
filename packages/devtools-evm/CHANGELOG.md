# @layerzerolabs/devtools-evm

## 0.3.10

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies
- Updated dependencies [8b32a79]
  - @layerzerolabs/devtools@0.3.15
  - @layerzerolabs/io-devtools@0.1.10

## 0.3.9

### Patch Changes

- 8935369: Add experimental support for batched sending
- c13cee5: Use getNextNonce when proposing transactions to gnosis
- Updated dependencies [8935369]
  - @layerzerolabs/devtools@0.3.13

## 0.3.8

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema
- Updated dependencies [938ac3d]
  - @layerzerolabs/io-devtools@0.1.9
  - @layerzerolabs/devtools@0.3.11

## 0.3.7

### Patch Changes

- f865318: Update @LayerZero-Labs dependencies to 2.3.3
- Updated dependencies [f865318]
  - @layerzerolabs/devtools@0.3.10

## 0.3.6

### Patch Changes

- b20ba18: Add ability to use named signers

## 0.3.5

### Patch Changes

- 0f853ed: Adding `AsyncRetriable` to SDKs
- Updated dependencies [0f853ed]
- Updated dependencies [958ae04]
  - @layerzerolabs/devtools@0.3.6

## 0.3.4

### Patch Changes

- 947a9aa: Add Configurator helper type
- Updated dependencies [947a9aa]
- Updated dependencies [947a9aa]
  - @layerzerolabs/devtools@0.3.5

## 0.3.3

### Patch Changes

- e16c864: Update @LayerZero-Labs dependencies to 2.1.27
- e16c864: Update executor config schema & types after update to 2.1.27
- Updated dependencies [e16c864]
- Updated dependencies [e16c864]
  - @layerzerolabs/devtools@0.3.4

## 0.3.2

### Patch Changes

- c09680a: Update @LayerZero-Labs dependencies to 2.1.25
- Updated dependencies [c09680a]
  - @layerzerolabs/devtools@0.3.3

## 0.3.1

### Patch Changes

- 059d817: Add the ability to choose gnosis safe for transaction signing; add the ability to pick signer index/address

## 0.3.0

### Minor Changes

- a4093ab: Adapt LayerZero package updates, including OApp version 2

### Patch Changes

- Updated dependencies [a4093ab]
  - @layerzerolabs/devtools@0.3.0

## 0.2.12

### Patch Changes

- f9987d3: Add retries, timeout and forkBlockNumber properties to AnvilOptions
- ff6427c: Align the dependency ranges
- ff6427c: Update @LayerZero-Labs dependencies to 2.1.18
- Updated dependencies [ff6427c]
- Updated dependencies [ff6427c]
  - @layerzerolabs/devtools@0.2.10

## 0.2.11

### Patch Changes

- 04e1e67: Add stateInterval to AnvilOptions

## 0.2.10

### Patch Changes

- 6d2e1f8: Add types for simulation
- Updated dependencies [6d2e1f8]
  - @layerzerolabs/devtools@0.2.9

## 0.2.9

### Patch Changes

- cc83f96: Add fp-ts to dependencies
- Updated dependencies [40c45c1]
- Updated dependencies [40c45c1]
- Updated dependencies [4292d81]
  - @layerzerolabs/devtools@0.2.8

## 0.2.8

### Patch Changes

- 54cf16e: Silence bigint-buffer warning
- Updated dependencies [54cf16e]
- Updated dependencies [2b9ae6a]
- Updated dependencies [2b9ae6a]
  - @layerzerolabs/devtools@0.2.7
  - @layerzerolabs/io-devtools@0.1.5

## 0.2.7

### Patch Changes

- e7ef1aa: Update @layerzero-labs dependencies to 2.1.15
- Updated dependencies [e7ef1aa]
  - @layerzerolabs/devtools@0.2.6

## 0.2.6

### Patch Changes

- b93a018: Update @layerzero-labs dependencies to 2.1.13
- Updated dependencies [b93a018]
  - @layerzerolabs/devtools@0.2.5

## 0.2.5

### Patch Changes

- 855fa36: Update to latest layerzerolabsmonorepo dependencies
- Updated dependencies [855fa36]
  - @layerzerolabs/devtools@0.2.4

## 0.2.4

### Patch Changes

- 2732dfc: Add support for submitting transactions to Gnosis Safes using @safe-global

## 0.2.3

### Patch Changes

- e0f41b5: Update @layerzero-labs dependencies
- Updated dependencies [e0f41b5]
  - @layerzerolabs/devtools@0.2.3

## 0.2.2

### Patch Changes

- 491b5a5: Upgrade contract Dependencies to 2.1.7
- Updated dependencies [491b5a5]
  - @layerzerolabs/devtools@0.2.2

## 0.2.1

### Patch Changes

- 71e355c: Update "@layerzerolabs/\*" dependencies
- Updated dependencies [71e355c]
  - @layerzerolabs/devtools@0.2.1

## 0.2.0

### Minor Changes

- 6e464f2: Specify EndpointV2 with the "V2" suffix wherever appropriate

### Patch Changes

- Updated dependencies [6e464f2]
  - @layerzerolabs/devtools@0.2.0

## 0.1.3

### Patch Changes

- 8227742: Update @layerzero-labs dependencies to 2.1.4
- Updated dependencies [8227742]
  - @layerzerolabs/devtools@0.1.4

## 0.1.2

### Patch Changes

- 5c58d69: Update @layerzerolabs dependencies to 2.1.3
- Updated dependencies [253c79e]
- Updated dependencies [5c58d69]
  - @layerzerolabs/io-devtools@0.1.2
  - @layerzerolabs/devtools@0.1.3

## 0.1.1

### Patch Changes

- f0036c5: Use monorepo 2.1.2 released dependencies
- Updated dependencies [f0036c5]
  - @layerzerolabs/devtools@0.1.1

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

### Patch Changes

- Updated dependencies [120adf1]
  - @layerzerolabs/io-devtools@0.1.0
  - @layerzerolabs/devtools@0.1.0

## 0.0.8

### Patch Changes

- fac1c06: Add addChecksum utility

## 0.0.7

### Patch Changes

- 8789236: Move bytes utilities to devtools
- Updated dependencies [8789236]
- Updated dependencies [9b4256a]
  - @layerzerolabs/devtools@0.0.6

## 0.0.6

### Patch Changes

- 0877186: Update @layerzerolabs dependencies to 2.0.25 and 2.0.26-rc1
- Updated dependencies [0877186]
  - @layerzerolabs/devtools@0.0.5

## 0.0.5

### Patch Changes

- 8931fb2: Add debug logging to SDKs

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies
- Updated dependencies [af8cc25]
  - @layerzerolabs/io-devtools@0.0.4
  - @layerzerolabs/devtools@0.0.4

## 0.0.3

### Patch Changes

- 5077bf1: Adjust the custom error messages - add better stringification
- Updated dependencies [5077bf1]
  - @layerzerolabs/io-devtools@0.0.3

## 0.0.2

### Patch Changes

- 70646b4: Add missing dependencies
- Updated dependencies [70646b4]
- Updated dependencies [b5991ca]
- Updated dependencies [2dac0da]
  - @layerzerolabs/devtools@0.0.2

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
- Updated dependencies [b74afbe]
  - @layerzerolabs/devtools@0.0.1
