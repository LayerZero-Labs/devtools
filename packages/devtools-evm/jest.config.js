/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
    cache: false,
    testEnvironment: 'node',
    setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
    moduleNameMapper: {
        '^@/(.*)$': '<rootDir>/src/$1',
    },
    transform: {
        '^.+\\.(t|j)sx?$': '@swc/jest',
    },
};
