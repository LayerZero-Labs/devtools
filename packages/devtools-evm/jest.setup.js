import jestExtended from 'jest-extended';

// add all jest-extended matchers
expect.extend(jestExtended);
