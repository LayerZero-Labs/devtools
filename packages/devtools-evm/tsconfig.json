{
  "extends": "../../tsconfig.json",
  "exclude": ["dist", "node_modules"],
  "include": ["src", "test", "*.config.ts"],
  "compilerOptions": {
    "types": ["node", "jest"],
    "paths": {
      "@/*": ["./src/*"]
    }
  }
}
