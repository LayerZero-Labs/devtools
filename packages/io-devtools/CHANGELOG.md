# @layerzerolabs/io-devtools

## 0.1.10

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies

## 0.1.9

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema

## 0.1.8

### Patch Changes

- 9af95ca: Add createWithAsyncLogger logging utility for async functions

## 0.1.7

### Patch Changes

- 8a4addf: Replace cli-table3 with table

## 0.1.6

### Patch Changes

- 2a36974: Add the ability to define configs as functions

## 0.1.5

### Patch Changes

- 54cf16e: Silence bigint-buffer warning

## 0.1.4

### Patch Changes

- 9e78abe: Update hardhat-deploy

## 0.1.3

### Patch Changes

- e766c86: Adding in Default Colors to print outs

## 0.1.2

### Patch Changes

- 253c79e: Rename description to hint for select prompts
- 5c58d69: Update @layerzerolabs dependencies to 2.1.3

## 0.1.1

### Patch Changes

- 40e8dba: Remove CommaSeparatedValuesSchema in favor of splitCommaSeparated; Move LogLevelSchema; Export isLogLevel

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

## 0.0.7

### Patch Changes

- 81d7eee: Add promptForText utility

## 0.0.6

### Patch Changes

- 2b36ed9: Updating Table for Get OApp Config Task
- 2b36ed9: Rafactor some of the printers in io-devtools

## 0.0.5

### Patch Changes

- ed450c4: Add promptToSelectOne and promptToSelectMultiple utilities

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies

## 0.0.3

### Patch Changes

- 5077bf1: Unmount the component when using printTable

## 0.0.2

### Patch Changes

- e0be8b7: Reexport Logger type
- 70646b4: Add missing dependencies

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
