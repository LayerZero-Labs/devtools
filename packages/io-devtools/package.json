{
  "name": "@layerzerolabs/io-devtools",
  "version": "0.1.10",
  "description": "Utilities for working with I/O in LayerZero devtools",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/devtools.git",
    "directory": "packages/io-devtools"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "types": "./dist/*/index.d.ts",
      "require": "./dist/*/index.js",
      "import": "./dist/*/index.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "swag"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "$npm_execpath tsup",
    "clean": "rm -rf dist",
    "dev": "$npm_execpath tsup --watch",
    "lint": "$npm_execpath eslint '**/*.{js,ts,json}'",
    "lint:fix": "eslint --fix '**/*.{js,ts,json}'",
    "test": "jest --ci"
  },
  "dependencies": {
    "chalk": "^4.1.2",
    "logform": "^2.6.0",
    "prompts": "^2.4.2",
    "table": "~6.8.2",
    "winston": "^3.11.0"
  },
  "devDependencies": {
    "@swc/core": "^1.4.0",
    "@swc/jest": "^0.2.36",
    "@types/ink-gradient": "^2.0.4",
    "@types/jest": "^29.5.12",
    "@types/prompts": "^2.4.9",
    "@types/react": "^17.0.75",
    "fast-check": "^3.15.1",
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "jest": "^29.7.0",
    "jest-extended": "^4.0.2",
    "react": "^17.0.2",
    "ts-node": "^10.9.2",
    "tslib": "~2.6.2",
    "tsup": "~8.0.1",
    "typescript": "^5.3.3",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "react": "^17.0.2",
    "yoga-layout-prebuilt": "^1.9.6",
    "zod": "^3.22.4"
  },
  "peerDependenciesMeta": {
    "ink": {
      "optional": true
    },
    "ink-gradient": {
      "optional": true
    },
    "ink-table": {
      "optional": true
    },
    "react": {
      "optional": true
    },
    "yoga-layout-prebuilt": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
