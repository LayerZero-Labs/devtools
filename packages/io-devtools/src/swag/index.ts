export * from './printer'
export * from './renderer'
