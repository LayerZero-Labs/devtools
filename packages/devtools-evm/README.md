<p align="center">
  <a href="https://layerzero.network">
    <img alt="LayerZero" style="max-width: 500px" src="https://d3a2dpnnrypp5h.cloudfront.net/bridge-app/lz.png"/>
  </a>
</p>

<h1 align="center">@layerzerolabs/devtools-evm</h1>

<!-- The badges section -->
<p align="center">
  <!-- Shields.io NPM published package version -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/devtools-evm"><img alt="NPM Version" src="https://img.shields.io/npm/v/@layerzerolabs/devtools-evm"/></a>
  <!-- Shields.io NPM downloads -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/devtools-evm"><img alt="Downloads" src="https://img.shields.io/npm/dm/@layerzerolabs/devtools-evm"/></a>
  <!-- Shields.io license badge -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/devtools-evm"><img alt="NPM License" src="https://img.shields.io/npm/l/@layerzerolabs/devtools-evm"/></a>
</p>

Utilities for working with LayerZero EVM contracts.

## Installation

```bash
yarn add @layerzerolabs/devtools-evm

pnpm add @layerzerolabs/devtools-evm

npm install @layerzerolabs/devtools-evm
```

### Address utilities

#### ignoreZero(address: Address | null | undefined)

Turns EVM zero addresses to `undefined`

```typescript
import { ignoreZero } from "@layerzerolabs/devtools-evm";

ignoreZero("0xEe6cF2E1Bc7645F8439d241ce37820305F2BB3F8"); // Returns '0xEe6cF2E1Bc7645F8439d241ce37820305F2BB3F8'
ignoreZero("0x0000000000000000000000000000000000000000"); // Returns undefined
ignoreZero(undefined); // Returns undefined
ignoreZero(null); // Returns undefined
```

#### makeZeroAddress(address)

Turns `null` and `undefined` into EVM zero address

```typescript
import { makeZeroAddress } from "@layerzerolabs/devtools-evm";

makeZeroAddress("0xEe6cF2E1Bc7645F8439d241ce37820305F2BB3F8"); // Returns '0xEe6cF2E1Bc7645F8439d241ce37820305F2BB3F8'
makeZeroAddress("0x0000000000000000000000000000000000000000"); // Returns '0x0000000000000000000000000000000000000000'
makeZeroAddress(undefined); // Returns '0x0000000000000000000000000000000000000000'
makeZeroAddress(null); // Returns '0x0000000000000000000000000000000000000000'
```
