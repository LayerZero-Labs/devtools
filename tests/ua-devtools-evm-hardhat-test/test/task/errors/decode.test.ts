import hre from 'hardhat'
import { TASK_COMPILE } from 'hardhat/builtin-tasks/task-names'
import { TASK_LZ_ERRORS_DECODE } from '@layerzerolabs/ua-devtools-evm-hardhat'
import { printRecord } from '@layerzerolabs/io-devtools/swag'
import { CustomError, PanicError, RevertError, UnknownError } from '@layerzerolabs/devtools-evm'

jest.mock('@layerzerolabs/io-devtools/swag', () => ({
    printRecord: jest.fn(),
    printLogo: jest.fn(),
}))

const printRecordMock = printRecord as jest.Mock
const runMock = jest.spyOn(hre, 'run')

describe(`task ${TASK_LZ_ERRORS_DECODE}`, () => {
    beforeEach(() => {
        printRecordMock.mockReset()
        runMock.mockClear()
    })

    it('should compile contracts', async () => {
        await hre.run(TASK_LZ_ERRORS_DECODE, { hash: '' })

        // For some reason even though we did not specify any arguments to the compile task,
        // jest still sees some aarguments being passed so we need to pass those to make this expect work
        expect(runMock).toHaveBeenCalledWith(TASK_COMPILE, undefined, {}, undefined)
    })

    it('should not print anything if the error cannot be decoded', async () => {
        const result = await hre.run(TASK_LZ_ERRORS_DECODE, { hash: '0x0' })

        expect(result).toBeInstanceOf(UnknownError)
        expect(printRecordMock).not.toHaveBeenCalled()
    })

    it('should print PanicError details', async () => {
        const result = await hre.run(TASK_LZ_ERRORS_DECODE, {
            hash: '0x4e487b710000000000000000000000000000000000000000000000000000000000000006',
        })

        expect(result).toBeInstanceOf(PanicError)
        expect(printRecordMock).toHaveBeenCalledTimes(1)
        expect(printRecordMock.mock.calls[0]).toMatchSnapshot()
    })

    it('should print RevertError details', async () => {
        const result = await hre.run(TASK_LZ_ERRORS_DECODE, {
            hash: '0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000066d79206261640000000000000000000000000000000000000000000000000000',
        })

        expect(result).toBeInstanceOf(RevertError)
        expect(printRecordMock).toHaveBeenCalledTimes(1)
        expect(printRecordMock.mock.calls[0]).toMatchSnapshot()
    })

    it('should print CustomError details', async () => {
        const result = await hre.run(TASK_LZ_ERRORS_DECODE, { hash: '0x9153304b' })

        expect(result).toBeInstanceOf(CustomError)
        expect(printRecordMock).toHaveBeenCalledTimes(1)
        expect(printRecordMock.mock.calls[0]).toMatchSnapshot()
    })

    it.each(['0x3a81d6fc', '0x81da67ee'])('should print CustomError details if the error is %s', async (hash) => {
        const result = await hre.run(TASK_LZ_ERRORS_DECODE, { hash })

        expect(result).toBeInstanceOf(CustomError)
        expect(printRecordMock).toHaveBeenCalledTimes(1)
        expect(printRecordMock.mock.calls[0]).toMatchSnapshot()
    })
})
