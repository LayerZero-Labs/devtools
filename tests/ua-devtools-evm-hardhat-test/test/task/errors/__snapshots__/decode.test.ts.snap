// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task lz:errors:decode should print CustomError details 1`] = `
[
  {
    "Error": "CustomError: Contract reverted with custom error. Error UnsupportedEid()",
    "Type": "Custom error",
  },
]
`;

exports[`task lz:errors:decode should print CustomError details if the error is 0x3a81d6fc 1`] = `
[
  {
    "Error": "CustomError: Contract reverted with custom error. Error AlreadyRegistered()",
    "Type": "Custom error",
  },
]
`;

exports[`task lz:errors:decode should print CustomError details if the error is 0x81da67ee 1`] = `
[
  {
    "Error": "CustomError: Contract reverted with custom error. Error InvalidOptionalDVNThreshold()",
    "Type": "Custom error",
  },
]
`;

exports[`task lz:errors:decode should print PanicError details 1`] = `
[
  {
    "Error": "PanicError: Contract panicked (assert() has been called). Error code 6",
    "Type": "Panic error",
  },
]
`;

exports[`task lz:errors:decode should print RevertError details 1`] = `
[
  {
    "Error": "RevertError: Contract reverted. Error reason 'my bad'",
    "Type": "Revert error",
  },
]
`;
